merkletree/proof_test.go

```diff
@@ -2,192 +2,145 @@ package merkletree
-	}
-	if err := m.Set(index3, key3, val3); err != nil {
-		t.Fatal(err)
-	}
-
-	m.recomputeHash()
-
-	ap1 := m.Get(index1)
-	if ap1.Leaf.Value == nil {
-		t.Error("Cannot find key:", key1)
-		return
-	}
-	ap2 := m.Get(index2)
-	if ap2.Leaf.Value == nil {
-		t.Error("Cannot find key:", key2)
-		return
-	}
-	ap3 := m.Get(index3)
-	if ap3.Leaf.Value == nil {
-		t.Error("Cannot find key:", key3)
-		return
-	}
+type mockTest struct {
+	key   string
+	value []byte
+	index []byte
+	want  ProofType
+}
 
-	// proof of inclusion
-	proof := m.Get(index3)
-	if proof.Leaf.Value == nil {
-		t.Fatal("Expect returned leaf's value is not nil")
-	}
-	// ensure this is a proof of inclusion by comparing the returned indices
-	// and verifying the VRF index as well.
-	if !bytes.Equal(proof.LookupIndex, proof.Leaf.Index) ||
-		!bytes.Equal(vrfPrivKey1.Compute([]byte(key3)), proof.LookupIndex) {
-		t.Fatal("Expect a proof of inclusion")
-	}
-	// verify auth path
-	if proof.Verify([]byte(key3), val3, m.hash) != nil {
-		t.Error("Proof of inclusion verification failed.")
-	}
+const letterBytes = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
+const (
+	letterIdxBits = 6                    // 6 bits to represent a letter index
+	letterIdxMask = 1<<letterIdxBits - 1 // All 1-bits, as many as letterIdxBits
+	letterIdxMax  = 63 / letterIdxBits   // # of letter indices fitting in 63 bits
+)
 
-	// proof of absence
-	absentIndex := vrfPrivKey1.Compute([]byte("123"))
-	proof = m.Get(absentIndex) // shares the same prefix with an empty node
-	if proof.Leaf.Value != nil {
-		t.Fatal("Expect returned leaf's value is nil")
-	}
-	// ensure this is a proof of absence
-	if bytes.Equal(proof.LookupIndex, proof.Leaf.Index) ||
-		!bytes.Equal(vrfPrivKey1.Compute([]byte("123")), proof.LookupIndex) {
-		t.Fatal("Expect a proof of absence")
-	}
-	if proof.Verify([]byte("123"), nil, m.hash) != nil {
-		t.Error("Proof of absence verification failed.")
-	}
+var src = rand.NewSource(time.Now().UnixNano())
+
+// Credit: https://stackoverflow.com/a/31832326
+func RandStringBytesMaskImprSrc(n int) string {
+	b := make([]byte, n)
+	// A src.Int63() generates 63 random bits, enough for letterIdxMax characters!
+	for i, cache, remain := n-1, src.Int63(), letterIdxMax; i >= 0; {
+		if remain == 0 {
+			cache, remain = src.Int63(), letterIdxMax
+		}
+		if idx := int(cache & letterIdxMask); idx < len(letterBytes) {
+			b[i] = letterBytes[idx]
+			i--
+		}
+		cache >>= letterIdxBits
+		remain--
+	}
+
+	return string(b)
 }
 
-func TestVerifyProofSamePrefix(t *testing.T) {
-	m, err := NewMerkleTree()
-	if err != nil {
-		t.Fatal(err)
-	}
-
-	key1 := "key1"
-	index1 := vrfPrivKey1.Compute([]byte(key1))
-	val1 := []byte("value1")
-	if err := m.Set(index1, key1, val1); err != nil {
-		t.Fatal(err)
-	}
+var N uint64 = 3 // number of inclusions.
+
+func setup(t *testing.T) (*MerkleTree, []*mockTest) {
+	m := newTestTree(t)
+
+	tuple := []*mockTest{}
+	for i := uint64(0); i < uint64(N); i++ {
```

Comment:
Insertion and lookups depends on test cases, so IMO we don't really need to refactor them into smaller functions.

Reply:
That makes sense. But there still is a lot of repeated test structure, and that's what I'm suggesting we refactor. It will make the tests cleaner and a lot easier to write new ones. I think it would be more valuable and helpful to us and other developers in the future if we could abstract away most of the internals of testing and just had to pass in the test cases we want to test. If it doesn't take too much effort, we should take care of that in this PR.